jax/_src/scipy/fft.py

```diff
@@ -0,0 +1,76 @@
+# Copyright 2020 Google LLC
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     https://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+import scipy.fft as osp_fft
+from jax import lax, numpy as jnp, partial
+from jax._src.lax.lax import _canonicalize_axis
+from jax._src.numpy.util import _wraps
+
+def _W4(N, k):
+  return jnp.exp(-.5j * jnp.pi * k / N)
+
+def _dct_interleave(x, axis):
+  v0 = lax.slice_in_dim(x, None, None, 2, axis)
+  v1 = lax.rev(lax.slice_in_dim(x, 1, None, 2, axis), (axis,))
+  return lax.concatenate([v0, v1], axis)
+
+# Implementation based on
+# John Makhoul: A Fast Cosine Transform in One and Two Dimensions (1980)
+
+@_wraps(osp_fft.dct)
+def dct(x, type=2, n=None, axis=-1, norm=None):
+  if type != 2:
+    raise NotImplementedError('Only DCT type 2 is implemented.')
+  if n is not None or norm is not None:
+    raise NotImplementedError('DCT is only implemented for n=None, norm=None.')
+
+  axis = _canonicalize_axis(axis, x.ndim)
+  N = x.shape[axis]
+  v = _dct_interleave(x, axis)
+  V = jnp.fft.fft(v, axis=axis)
+  k = lax.reshape(jnp.arange(N), [N if a == axis else 1 for a in range(x.ndim)])
```

Comment:
Prefer `lax.expand_dims` over `lax.reshape` for this (see note in `lax.reshape` docstring). I think something like this is equivalent:
```python
k1 = lax.expand_dims(jnp.arange(N), [a for a in range(x.ndim) if a != axis1])
```
Same elsewhere